src/go/cmd/cache.go

```diff
@@ -0,0 +1,162 @@
+func addModuleToCache(ctx context.Context, mod module.Version) error {
+	escaped, err := module.EscapePath(mod.Path)
+	if err != nil {
+		return fmt.Errorf("unescapeable module path %q: %w", mod.Path, err)
+	}
+	u, err := url.Parse("https://" + path.Join("proxy.golang.org", escaped, "@v", mod.Version+".zip"))
+	if err != nil {
+		return fmt.Errorf("failed to parse module URL: %w", err)
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+	if err != nil {
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to download module zip from %s: %w", u, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		msg := "module proxy responded %d"
+		if resp.Body != nil {
+			b, _ := io.ReadAll(resp.Body)
+			msg += ": " + string(b)
+		}
+		return fmt.Errorf(msg, resp.StatusCode)
+	}
+	zp := cachePathForZip(mod)
+	d := filepath.Dir(zp)
+	err = os.MkdirAll(d, 0700)
+	if err != nil {
+		return fmt.Errorf("failed to create %s: %w", d, err)
+	}
+	f, err := os.Create(zp)
+	if err != nil {
+		return fmt.Errorf("failed to create %s: %w", zp, err)
+	}
+	defer f.Close()
+	_, err = io.Copy(f, resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to write %s: %w", zp, err)
+	}
+	err = zip.Unzip(cachePathForSource(mod), mod, zp)
+	if err != nil {
+		return fmt.Errorf("failed to unzip %s: %w", zp, err)
+	}
+	return nil
+}
+
+// cachedModule returns a Module for mod if it's in either the local Go mod
+// cache or apiviewgo cache. It returns errCachedModuleNotFound when the
+// module isn't in either cache.
+func cachedModule(mod module.Version) (*Module, error) {
+	if modCache := os.Getenv("GOMODCACHE"); modCache != "" {
+		d := filepath.Join(modCache, mustEscape(mod.Path)) + "@" + mod.Version
+		if _, err := os.Stat(filepath.Join(d, "go.mod")); err == nil {
+			return NewModule(d)
+		}
+	}
+	d := cachePathForSource(mod)
+	if _, err := os.Stat(filepath.Join(d, "go.mod")); err == nil {
+		return NewModule(d)
+	}
+	zp := cachePathForZip(mod)
+	if _, err := os.Stat(zp); err == nil {
+		return modFromZip(mod)
+	}
+	return nil, errCachedModuleNotFound
+}
+
+// mustEscape escapes modPath. It panics when that fails.
+func mustEscape(modPath string) string {
+	escaped, err := module.EscapePath(modPath)
+	if err != nil {
+		panic(fmt.Errorf("failed to escape module path: %w", err))
+	}
+	return escaped
+}
+
+// modFromZip unzips the module at mod's zip file in the local cache
+func modFromZip(mod module.Version) (*Module, error) {
+	d := cachePathForSource(mod)
+	err := zip.Unzip(d, mod, cachePathForZip(mod))
+	if err != nil {
+		return nil, fmt.Errorf("failed to unzip module: %w", err)
+	}
+	if _, err := os.Stat(filepath.Join(d, "go.mod")); err != nil {
+		return nil, fmt.Errorf("go.mod not found in %s", d)
+	}
+	return NewModule(d)
+}
+
+// the apiviewgo cache looks like:
+//     /tmp/apiviewgo/cache
```

Comment:
Is this temp folder version resilient? What happens if two reviews are being generated, will they stomp on each other to create the cache?

Should we make the cache tied to the review?

Reply:
Good question. I wrote this on the assumption that apiviewgo runs sequentially but not concurrently. I don't actually know the architecture of our automation though, so I'll have to ask someone about the details. Concurrent processes would race for the same module version, possibly causing one to fail or produce garbage due to reading torn files.

Reply:
https://pkg.go.dev/os#MkdirTemp will help you here.

Reply:
`MkdirTemp` would make the cache ephemeral. I was reluctant to do that because I imagine this mechanism becoming the solution for #8698, however it's a small efficiency given we have 3 cross-module aliases in the SDK today. So, I'll make the change; we can revisit the synchronization question later, if durable caching becomes interesting.